File: app/src/main/java/marabillas/loremar/lmvideodownloader/browsing_feature/VideoDetailsFetcher.java
/*
 *     LM videodownloader is a browser app for android, made to easily
 *     download videos.
 *     Copyright (C) 2018 Loremar Marabillas
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package marabillas.loremar.lmvideodownloader.browsing_feature;

import java.text.DecimalFormat;
import java.util.Locale;

import wseemann.media.FFmpegMediaMetadataRetriever;

public final class VideoDetailsFetcher {
    private FFmpegMediaMetadataRetriever metadataRetriever = new FFmpegMediaMetadataRetriever();

    public void fetchDetails(final String url, final FetchDetailsListener listener) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    metadataRetriever.setDataSource(url);
                    String filename = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_FILENAME);
                    String title = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_TITLE);
                    String vcodec = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_VIDEO_CODEC);
                    String acodec = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_AUDIO_CODEC);
                    String duration = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_DURATION);
                    String filesize = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_FILESIZE);
                    String width = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH);
                    String height = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT);
                    String bitrate = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_VARIANT_BITRATE);
                    String framerate = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_FRAMERATE);
                    String encoder = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_ENCODER);
                    String encodedBy = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_ENCODED_BY);
                    String date = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_DATE);
                    String creationTime = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_CREATION_TIME);
                    String artist = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_ARTIST);
                    String album = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_ALBUM);
                    String albumArtist = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_ALBUM_ARTIST);
                    String track = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_TRACK);
                    String genre = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_GENRE);
                    String composer = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_COMPOSER);
                    String performaer = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_PERFORMER);
                    String copyright = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_COPYRIGHT);
                    String publisher = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_PUBLISHER);
                    String language = metadataRetriever.extractMetadata(FFmpegMediaMetadataRetriever.METADATA_KEY_LANGUAGE);

                    StringBuilder sb = new StringBuilder();
                    if (filename != null) sb.append("\nFile: ").append(filename);
                    if (title != null) sb.append("\nTitle: ").append(title);
                    sb.append("\nVideo Codec: ").append((vcodec != null) ? vcodec : "none");
                    sb.append("\nAudio Codec: ").append((acodec != null) ? acodec : "none");
                    String durationFormatted = formatDuration(duration);
                    if (durationFormatted != null)
                        sb.append("\nDuration: ").append(durationFormatted);
                    String filesizeFormatted = formatFilesize(filesize);
                    if (filesizeFormatted != null)
                        sb.append("\nFilesize: ").append(filesizeFormatted);
                    if (width != null) sb.append("\nWidth: ").append(width);
                    if (height != null) sb.append("\nHeight: ").append(height);
                    if (bitrate != null) sb.append("\nBitrate: ").append(bitrate);
                    if (framerate != null)
                        sb.append("\nFramerate: ").append(framerate).append(" fps");
                    if (encoder != null) sb.append("\nEncoder: ").append(encoder);
                    if (encodedBy != null) sb.append("\nEncoded By: ").append(encodedBy);
                    if (date != null) sb.append("\nDate: ").append(date);
                    if (creationTime != null) sb.append("\nCreation Time: ").append(creationTime);
                    if (artist != null) sb.append("\nArtist: ").append(artist);
                    if (album != null) sb.append("\nAlbum: ").append(album);
                    if (albumArtist != null) sb.append("\nAlbum Artist: ").append(albumArtist);
                    if (track != null) sb.append("\nTrack: ").append(track);
                    if (genre != null) sb.append("\nGenre: ").append(genre);
                    if (composer != null) sb.append("\nComposer: ").append(composer);
                    if (performaer != null) sb.append("\nPerformer: ").append(performaer);
                    if (copyright != null) sb.append("\nCopyright: ").append(copyright);
                    if (publisher != null) sb.append("\nPublisher: ").append(publisher);
                    if (language != null) sb.append("\nLanguage: ").append(language);

                    if (sb.length() > 0) {
                        sb.deleteCharAt(0);
                    }

                    listener.onFetched(sb.toString());
                } catch (IllegalArgumentException e) {
                    listener.onUnFetched(e.getMessage());
                }
            }
        }).start();
    }

    private String formatDuration(String duration) {
        try {
            long totalSecs = Long.parseLong(duration) / 1000;
            long mils = totalSecs % 1000;
            long s = totalSecs % 60;
            long m = (totalSecs / 60) % 60;
            long h = (totalSecs / (60 * 60)) % 24;
            return String.format(Locale.US, "%02d:%02d:%02d.%d", h, m, s, mils);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    private String formatFilesize(String filesize) {
        try {
            long size = Long.parseLong(filesize);
            if (size <= 0) return "0";
            final String[] units = new String[]{"B", "kB", "MB", "GB", "TB"};
            int digitGroups = (int) (Math.log10(size) / Math.log10(1024));
            return new DecimalFormat("#,##0.#").format(size / Math.pow(1024, digitGroups))
                    + " " + units[digitGroups];
        } catch (NumberFormatException e) {
            return null;
        }
    }

    void close() {
        metadataRetriever.release();
    }

    interface FetchDetailsListener {
        void onUnFetched(String message);

        void onFetched(String details);
    }
}

File: base/src/main/java/marabillas/loremar/beedio/base/media/VideoDetailsFetcher.kt
/*
 * Beedio is an Android app for downloading videos
 * Copyright (C) 2019 Loremar Marabillas
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package marabillas.loremar.beedio.base.media

import android.graphics.Bitmap
import wseemann.media.FFmpegMediaMetadataRetriever
import wseemann.media.FFmpegMediaMetadataRetriever.*
import java.text.DecimalFormat
import java.util.*
import kotlin.math.log10
import kotlin.math.pow


class VideoDetailsFetcher {
    private val metadataRetriever = FFmpegMediaMetadataRetriever()
    private var isCancelled = false

    fun fetchDetails(url: String, fetchListener: FetchListener) {
        isCancelled = false
        try {
            metadataRetriever.setDataSource(url)
            val details = VideoDetails(
                    filename = extract(METADATA_KEY_FILENAME),
                    title = extract(METADATA_KEY_TITLE),
                    vcodec = extract(METADATA_KEY_VIDEO_CODEC),
                    acodec = extract(METADATA_KEY_AUDIO_CODEC),
                    duration = extract(METADATA_KEY_DURATION)?.formatDuration(),
                    filesize = extract(METADATA_KEY_FILESIZE)?.formatFilesize(),
                    width = extract(METADATA_KEY_VIDEO_WIDTH),
                    height = extract(METADATA_KEY_VIDEO_HEIGHT),
                    bitrate = extract(METADATA_KEY_VARIANT_BITRATE),
                    framerate = extract(METADATA_KEY_FRAMERATE),
                    encoder = extract(METADATA_KEY_ENCODER),
                    encodedBy = extract(METADATA_KEY_ENCODED_BY),
                    date = extract(METADATA_KEY_DATE),
                    creationTime = extract(METADATA_KEY_CREATION_TIME),
                    artist = extract(METADATA_KEY_ARTIST),
                    album = extract(METADATA_KEY_ALBUM),
                    albumArtist = extract(METADATA_KEY_ALBUM_ARTIST),
                    track = extract(METADATA_KEY_TRACK),
                    genre = extract(METADATA_KEY_GENRE),
                    composer = extract(METADATA_KEY_COMPOSER),
                    performer = extract(METADATA_KEY_PERFORMER),
                    copyright = extract(METADATA_KEY_COPYRIGHT),
                    publisher = extract(METADATA_KEY_PUBLISHER),
                    language = extract(METADATA_KEY_LANGUAGE)
            )
            if (details.vcodec != null)
                details.thumbnail = metadataRetriever.frameAtTime
            fetchListener.onFetched(details)
        } catch (e: CancelledException) {
            fetchListener.onUnFetched(e)
        } catch (e: IllegalArgumentException) {
            fetchListener.onUnFetched(e)
        } catch (e: Exception) {
            fetchListener.onUnFetched(e)
        }
    }

    fun fetchMiniDetails(url: String, fetchListener: FetchListener) {
        isCancelled = false
        try {
            metadataRetriever.setDataSource(url)
            val details = VideoDetails(
                    filename = extract(METADATA_KEY_FILENAME),
                    title = extract(METADATA_KEY_TITLE),
                    vcodec = extract(METADATA_KEY_VIDEO_CODEC),
                    duration = extract(METADATA_KEY_DURATION)?.formatDurationShort()
            )
            if (details.vcodec != null)
                details.thumbnail = metadataRetriever.frameAtTime
            fetchListener.onFetched(details)
        } catch (e: CancelledException) {
            fetchListener.onUnFetched(e)
        } catch (e: IllegalArgumentException) {
            fetchListener.onUnFetched(e)
        } catch (e: Exception) {
            fetchListener.onUnFetched(e)
        }
    }

    private fun extract(key: String) = if (!isCancelled) metadataRetriever.extractMetadata(key)
    else throw CancelledException()

    fun cancel() {
        isCancelled = true
    }

    private fun String.formatDuration(): String? {
        return try {
            val totalSecs = toLong() / 1000
            val mils = totalSecs % 1000
            val s = totalSecs % 60
            val m = totalSecs / 60 % 60
            val h = totalSecs / (60 * 60) % 24
            String.format(Locale.US, "%02d:%02d:%02d.%d", h, m, s, mils)
        } catch (e: NumberFormatException) {
            null
        }
    }

    private fun String.formatDurationShort(): String? {
        return try {
            val totalSecs = toLong() / 1000
            val s = totalSecs % 60
            val m = totalSecs / 60 % 60
            val h = totalSecs / (60 * 60) % 24
            when {
                h > 0 -> String.format(Locale.US, "%d:%02d:%02d", h, m, s)
                m > 0 -> String.format(Locale.US, "%d:%02d", m, s)
                else -> String.format(Locale.US, "%d seconds", s)
            }
        } catch (e: java.lang.NumberFormatException) {
            null
        }
    }

    private fun String.formatFilesize(): String? {
        return try {
            val size: Long = toLong()
            if (size <= 0) return "0"
            val units = arrayOf("B", "kB", "MB", "GB", "TB")
            val digitGroups = (log10(size.toDouble()) / log10(1024.0)).toInt()
            DecimalFormat("#,##0.#").format(size / 1024.0.pow(digitGroups.toDouble()))
                    .toString() + " " + units[digitGroups]
        } catch (e: NumberFormatException) {
            null
        }
    }

    fun close() = metadataRetriever.release()

    interface FetchListener {
        fun onUnFetched(error: Throwable)
        fun onFetched(details: VideoDetails)
    }

    inner class CancelledException : Exception()
}

data class VideoDetails(
        val filename: String? = null,
        val title: String? = null,
        val vcodec: String? = null,
        var acodec: String? = null,
        val duration: String? = null,
        var filesize: String? = null,
        val width: String? = null,
        val height: String? = null,
        val bitrate: String? = null,
        val framerate: String? = null,
        val encoder: String? = null,
        val encodedBy: String? = null,
        val date: String? = null,
        val creationTime: String? = null,
        val artist: String? = null,
        val album: String? = null,
        val albumArtist: String? = null,
        val track: String? = null,
        val genre: String? = null,
        val composer: String? = null,
        val performer: String? = null,
        val copyright: String? = null,
        val publisher: String? = null,
        val language: String? = null,
        var thumbnail: Bitmap? = null
)